UI Texture 9 slice

ManevilleF opened the pull request:
> Follow up to #10588 

Enable Texture slicing for the following UI nodes:
- `ImageBundle`
- `ButtonBundle`

<img width="817" alt="Screenshot 2024-01-29 at 12 50 08" src="https://github.com/bevyengine/bevy/assets/26703856/dacae7ac-654f-42e9-bcd3-e8e653124306">

ManevilleF commented:
@alice-i-cecile I think this should go to 0.13 as long as the sprite slices, the UI feature is probably even more useful than the sprite counterpart

alice-i-cecile commented:
I agree: the UI equivalent is heavily awaited and I'd like to be able to land both at once.

alice-i-cecile commented on crates/bevy_sprite/src/texture_slice/slicer.rs:
```diff
@@ -198,12 +200,20 @@ impl TextureSlicer {
+            || self.border.top >= rect_size.y
+            || self.border.bottom >= rect_size.y
+        {
+            panic!("TextureSlicer border has out of bounds values");
```
Can we just return an empty vec or an Option instead of panicking?

ManevilleF replied:
We can but then it will silently fail without any slicing applied. So at least a `log::error!` message ?

alice-i-cecile replied:
Yeah, log an error and fail IMO.

alice-i-cecile commented on examples/ui/ui_texture_slice.rs:
```diff
@@ -0,0 +1,87 @@
+//! This example illustrates how to create a button that has its image sliced
```
This could use more meat. Why might a user want to do this?

ManevilleF replied:
I updated the docs, is that enough ?

alice-i-cecile replied:
Yep this is fine.

ManevilleF commented on crates/bevy_ui/src/texture_slice.rs:
```diff
@@ -0,0 +1,177 @@
+// This module is mostly copied and pasted from `bevy_sprite::texture_slice`
+//
+// A more centralized solution should be investigated in the future
```
Tbh most of the code is not duplicated the main components and logic are reused, what is *duplicated* and *adapted* is the `ComputedTextureSlices` code, so only the cache elements with the sprite/ui specific logic

ManevilleF replied:
Anyway in the future we might want to implement slicing directly in the shader

ManevilleF commented:
@alice-i-cecile What about the new textures ? Do you think keeping them makes sense or should I only keep the one in use ?

alice-i-cecile commented:
I like the new textures a lot: I think they clearly demonstrate the value in a more appropriate way.

ManevilleF commented:
> I like the new textures a lot: I think they clearly demonstrate the value in a more appropriate way.

Great, but I added the whole pack in the `assets` folder, so we might want to remove the unused ones from the version control

alice-i-cecile commented:
Yeah, let's strip it down to only the ones that we use. We can always add more later if we find a use in other examples.

ManevilleF commented on crates/bevy_sprite/src/texture_slice/slicer.rs:
```diff
@@ -198,13 +200,28 @@ impl TextureSlicer {
     ///
     /// * `rect` - The section of the texture to slice in 9 parts
     /// * `render_size` - The optional draw size of the texture. If not set the `rect` size will be used.
+    ///
+    /// # Panics
+    ///
+    /// Panics if any border values are bigger than `rect` half extents
```
oh right

ManevilleF commented on crates/bevy_sprite/src/texture_slice/slicer.rs:
```diff
@@ -198,13 +200,28 @@ impl TextureSlicer {
     ///
     /// * `rect` - The section of the texture to slice in 9 parts
     /// * `render_size` - The optional draw size of the texture. If not set the `rect` size will be used.
+    ///
+    /// # Panics
+    ///
+    /// Panics if any border values are bigger than `rect` half extents
```
```suggestion
```

ManevilleF commented on crates/bevy_sprite/src/texture_slice/slicer.rs:
```diff
@@ -1,3 +1,5 @@
+use core::panic;
+
```
```suggestion
```

odecay commented:
Working on giving this a review, I noticed some left over unused system set ordering which seems like it was changed here https://github.com/bevyengine/bevy/pull/5103#discussion_r1480569178 in the Texture Atlas rework.

mockersf commented:
could you also add the assets you're adding to https://github.com/bevyengine/bevy/blob/main/CREDITS.md with their direct link and their license?

mockersf commented on crates/bevy_sprite/src/lib.rs:
```diff
@@ -137,7 +136,7 @@ pub fn calculate_bounds_2d(
         (Entity, &Sprite, &Handle<Image>, Option<&TextureAtlas>),
         (
-            Or<(Without<Aabb>, Changed<Sprite>)>,
+            Or<(Without<Aabb>, Changed<Sprite>, Changed<TextureAtlas>)>,
```
wouldn't that make the `TextureAtlas` component mandatory?

ManevilleF replied:
I don't think so, it's in a `Or` filter

mockersf commented:
it's a CC0 license so the license file is not strictly needed, it could be removed.

odecay commented on crates/bevy_ui/src/texture_slice.rs:
```diff
@@ -0,0 +1,177 @@
+                // TODO: support flip
+                flip_x: false,
+                // TODO: support flip
+                flip_y: true,
```
~~I tested to make sure with a modified non symmetrical slice image and we need to not flip the Y to get the expected result.~~

I tested both `flip_y: true` and `flip_y: false` and neither have an effect, but seems like with a modified non symmetrical image the result is flipped on y

odecay replied:
here is an example of above flip issue.
this is a non symmetrical slice image (note the placement of black spots along the left and bottom of image)
![panel-border-010_flip](https://github.com/bevyengine/bevy/assets/1379590/6e3193bf-4a97-49ab-a055-2bf9854bdd77)
this is the result when this image is used in the ui_texture_slice example (black spot on the bottom is flipped to be on top edge)
![flip_ui_9patch](https://github.com/bevyengine/bevy/assets/1379590/98c7cada-effd-49ea-8ac6-34b3525c3970)

ManevilleF replied:
Thanks for noticing, I'm fixing it

ManevilleF replied:
Should be fixed @odecay , it seems UI uses inversed Y coordinates than sprites

ManevilleF replied:
Your suggestion also allowed me to fix flipping for sprites

odecay reviewed (commented):
I see there is a note to support flipping, but it seems like there is an issue with the default orientation.

ManevilleF commented:
> I see there is a note to support flipping, but it seems like there is an issue with the default orientation.

It was tricky but I fixed the flipped issue and added flipping support for sliced textures

ManevilleF commented:
@alice-i-cecile D you know why the CI fails on `check-bans` ? What can I do to fix this ?

mockersf commented:
> @alice-i-cecile Do you know why the CI fails on `check-bans` ? What can I do to fix this ?

you can't, and it's not blocking for merging (it doesn't have the "required" tag)

mockersf reviewed (approved):
I'm unhappy about the need for a `compute_slices_on_asset_event` system, and even worse that it was already added for sprites.

I think 9-slice enabled ui elements and sprites should not be part of the existing bundles but in new bundles, or a "add this component `ImageScaleMode` to any entity with an image and it will be sliced/scaled/tiled, otherwise it will be just scaled and nothing fancy will happen"

This has already been added for sprites, so approving here for feature parity

odecay reviewed (approved):
Nice, the flipping works as expected now and the default orientation is correct. 
The ordering cleanup also looks :+1: 
LGTM

ManevilleF commented:
> 
> Yup, what I'm unhappy about is that any change to an image asset will go through all image (2d or UI) entities, even when you're not using 9 slice at all
> 

Honestly we can just remove those systems, the slices won't refresh if you change the image asset directly but users can be responsible to trigger change detection 


> > > I think 9-slice enabled ui elements and sprites should not be part of the existing bundles but in new bundles, or a "add this component `ImageScaleMode` to any entity with an image and it will be sliced/scaled/tiled, otherwise it will be just scaled and nothing fancy will happen"
> > 
> > Since the sprite slicing was not released yet we can remove the scale mode from the sprite and UI bundles, note that in UI it was only added to bundles with a UI image
> 
> I would like that. Do you think we should duplicate all bundles with an added `ImageScaleMode` component? Or have it as a free standing component, and show in the examples and doc "add this component and stuff happens"? @alice-i-cecile, do you have an opinion?

I feel we might be adding too many bundles that are basically the same as others with one extra component.
For example the sprite sheet bundle is strictly identical to the sprite bundle with one extra component. So maybe we could have some "basic" bundles and explain the possible additional behaviour through extra components? This would be fixed by optional components in bundles too I think
